### PR TITLE
View and set backend settings

### PR DESCRIPTION
**Pull request by sadatay**
As it stood, the value for the RMAPI customer ID and API key were per
deployment. Now however, it can be configured via the
/eholdings/configuration endpoint so that each tenant can maintain
their own set of credentials. We needed a way to detect if the
eholdings app is usable, and if its backend is configured
correctly, and also to be able to change credentials.

This change checks on every eholdings application entry to see if
infact there _is_ a  backend for eholdings and gives you instruction
on how to download it. If there is a backend, it will ensure that it
is configured properly, and if not, it will direct the user to a new
settings screen with which a tenant admin can view the current
configuration and make changes where necessary.

In order to support development against server configurations which
are hard to arrive at (such as a borked backend) the ability to start
the app in different scenarios was added. For example to start the app
without a backend, you would invoke the command:

```
yarn start --mirage no-backend
```

### CODE REVIEW

File: mirage/scenarios/unconfigured-backend.js [hunk 1]
```diff
@@ -0,0 +1,6 @@
+export default function noBackendScenario(server) {
```
Comment (wwilsman) [inline mirage/scenarios/unconfigured-backend.js]: This should be `unconfiguredBackend`

File: src/redux/request.js [hunk 2]
```diff
@@ -241,14 +241,15 @@ export function createRequestEpic({
...
+        let state = getState();
+        let { okapi } = state;
         let { method = 'GET' } = options;
         let body, headers = { 'X-Okapi-Tenant': okapi.tenant };
         let url = endpoint;
 
         // endpoint url
         if (typeof endpoint === 'function') {
-          url = `${okapi.url}/${endpoint(data, options)}`;
+          url = `${okapi.url}/${endpoint(data, options, state)}`;
```
Comment (wwilsman) [inline src/redux/request.js]: This change looks left over since it's unused

File: src/routes/application.js [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  renderApplication() {
+    let { children, application } = this.props;
+
+    if (application.isPending) {
+      return (<div>...loading</div>);
+    } else if (application.isResolved) {
+      let config = application.content;
+
+      return config.isValid ? children : this.renderNotValidConfig();
+    }
```
Comment (wwilsman) [inline src/routes/application.js]: We need to handle `isRejected` as well

```js
if (application.isPending) {
  return (<div>...loading</div>);
} else if (application.isRejected) {
  return this.renderNotValidConfig(); // or this.renderConfigError()
} else if (application.isResolved) {
  return !application.content.isValid
    ? this.renderNotValidConfig()
    : children;
}
```

File: src/routes/application.js [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+
+  renderNotValidConfig() {
+    return (
+      <div className="eholdings-unconfigured-backend" data-test-eholdings-unconfigured-backend>
+        <h1>Error: Knowledge Base Not Configured</h1>
+        <p>The E-Holdings application detected the presence of a knowledge base, but the knowledge base does not appear to be configured.  Please configure the knowledge base with your credentials on the settings page..</p>
+      </div>
+    );
+  }
+}
```
Comment (wwilsman) [inline src/routes/application.js]: Can we get some screenshots of these in the PR? 😉

File: tests/backend-configuration-test.js [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+        });
+
+      });
+    });
+
+    describe('configuring the mod-go-kb backend', function() {
+      it('has yet to be implemented');
+    });
+  }
+});
```
Comment (wwilsman) [inline tests/backend-configuration-test.js]: We should also add a test + scenario for the backend sending an error response (handling `.isRejected`).

File: tests/helpers.js [hunk 6]
```diff
@@ -5,8 +5,10 @@ import chaiJquery from 'chai-jquery';
 import $ from 'jquery';
 import { render, unmountComponentAtNode } from 'react-dom';
 import { convergeOn } from './it-will';
+import startMirage from '../mirage/start';
 import TestHarness from './harness';
 
+
```
Comment (wwilsman) [inline tests/helpers.js]: ✂️

File: src/redux/request.js [hunk 7]
```diff
@@ -241,7 +241,8 @@ export function createRequestEpic({
     action$.ofType(REQUEST_MAKE)
       .filter((action) => action.name === name)
       .switchMap(({ data = {}, options = {} }) => {
-        let { okapi } = getState();
+        let state = getState();
+        let { okapi } = state;
```
Comment (wwilsman) [inline src/redux/request.js]: This is unnecessary as well now. You can just ignore the changes in this whole file

**Comment by cherewaty** [comment]
Blocked by https://github.com/thefrontside/mod-kb-ebsco/pull/11

**Comment by cherewaty** [comment]
LtR <-> RtL

http://g.recordit.co/YBPxmMOARv.gif

File: mirage/config.js [hunk 8]
```diff
@@ -1,11 +1,31 @@
...
+  }]);
+
+  this.get('/_/proxy/modules/mod-kb-ebsco', {
+    id: 'mod-kb-ebsco',
+    name: 'kb-ebsco',
+    provides: [{
+      id: 'eholdings',
+      version: '0.1.0',
+      handlers: [{
+        methods: ['GET', 'POST', 'PUT', 'PATCH', 'DELETE'],
```
Comment (sadatay) [inline mirage/config.js]: this isn't quite accurate, but should be fine for how we're using it

**Pull request merged**